### PR TITLE
Add option for current product as a reference

### LINKED ISSUE #50916: Product Collection - Product Context inspector control for collections using product reference
**Introduction**

There is PR #50164, which introduces a product picker in Editor when a collection is added which requires a product context but it doesn't exist in the current template/page/post etc. This way, the merchant can manually provide a product context to the collection.

There is another PR #44878, which introduces Inspector control to change the selected product. This way, the merchant can change the selected product using the Inspector control.

**Scope of this task**

#44878 is only for context where product context doesn't exist. But following collections:
- Related Products
- Upsells
- Cross-sells

should have following UI in Product Context:
image
image

As you can see in above screenshots, we need control with following radio buttons:
- From the current product
- From a specific product

In scope of this task, we need to implement this control which should be shown when:
- `usesReference` is set to `product` i.e. `usesReference: ['product']` (See PR #49796 for more info)
- And collection is inside Product Context, for example inside:
  - Single Product template
  - Specific Product template
  - Single Product block etc.

### PR DESCRIPTION
**Pull request by kmanijak**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds an option to set a "current product" as a reference for a collection.

Recently, there was a feature added that merchants can manually choose a product which is a reference for the collection. However, when there's a reference provided out of the box (e.g. Single Product template or Single Product block), merchants should have the option to choose the current product as a reference.

This happens through radio control allowing 

From the current product | From a specific product - placeholder | From a specific product
-- | -- | --
image | image | image

Closes https://github.com/woocommerce/woocommerce/issues/50916

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Pre-requisite:
Install this plugin - [register-product-collection-tester.zip](https://github.com/user-attachments/files/17147190/register-product-collection-tester.zip)

#### When Product reference is available

1. Go to Single Product template
2. Add Product Collection - Product Context block
3. Make sure the "From current product" is chosen by default

image

5. Switch to "From specific product"
6. Choose some product
7. Switch back to "From current product" and back to "From specific product"
8. Expected: The same product is chosen as before, not placeholder

#### When Product reference is NOT available

1. Create new post
2. Add Product Collection - Product Context block
3. You're prompted to choose product reference, choose any
4. Make sure there's no radio button and there's only product picker in inspector controls

image

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/assets/images/blocks/product-collection/placeholder.svg
Comment (imanish003) [inline plugins/woocommerce/assets/images/blocks/product-collection/placeholder.svg]: I really like how this type of icons are handled in [new Product Editor](https://github.com/woocommerce/woocommerce/blob/add/50916-product-reference-updated/packages/js/product-editor/src/images) (See [storybook](https://woocommerce.github.io/woocommerce/?path=/story/product-editor-images-cashregister--default)). I don't think we need to change it but just wanted to mention it 🙂

Although, we should consider formatting this SVG icon for improved readability 🙂
```suggestion
<svg width="36" height="36" viewBox="0 0 36 36" fill="none"
	xmlns="http://www.w3.org/2000/svg">
	<g clip-path="url(#clip0_5404_82493)">
		<rect width="36" height="36" rx="3" fill="#F6F7F7" />
		<g clip-path="url(#clip1_5404_82493)">
			<path
				d="M25.8564 9.08257C22.302 9.08257 20.2435 9.70406 19.0697 11.0879C19.8294 8.40974 21.2333 6.39913 23.7979 4.57812C20.7588 4.9705 18.9947 6.12525 18.1403 7.77772C17.2859 6.12657 15.5219 4.97182 12.4828 4.57812C15.0473 6.39913 16.4512 8.40842 17.211 11.0879C16.0371 9.70406 13.9787 9.08257 10.4243 9.08257C14.6609 11.0721 16.2816 12.6732 17.6619 15.5423H18.6175C19.9977 12.6732 21.6198 11.0721 25.855 9.08257H25.8564Z"
				fill="#BBBBBB" />
			<path
				d="M20.9742 40.3274C22.3527 39.8165 27.4074 38.6946 27.4074 33.7983C27.4074 30.8901 24.7772 28.6965 21.7913 28.1873C23.1198 27.7533 25.0597 25.8204 25.0597 23.6518C25.0597 20.858 22.5315 19.7433 20.4646 19.311C20.4646 18.6733 22.7622 15.3828 22.7622 15.3828H13.2643C13.2643 15.3828 15.5619 18.6733 15.5619 19.311C13.4931 19.7451 10.9667 20.858 10.9667 23.6518C10.9667 25.8204 12.9067 27.7533 14.2352 28.1873C11.2474 28.6982 8.61905 30.8919 8.61905 33.7983C8.61905 38.6964 13.6737 39.8183 15.0523 40.3274C11.7838 40.6328 10.1496 41.9083 9.89568 42.9283H26.1325C25.8769 41.9083 24.2426 40.6328 20.9759 40.3274H20.9742Z"
				fill="#DDDDDD" />
		</g>
	</g>
	<defs>
		<clipPath id="clip0_5404_82493">
			<rect width="36" height="36" rx="3" fill="white" />
		</clipPath>
		<clipPath id="clip1_5404_82493">
			<rect width="38.3491" height="29.2493" fill="white"
				transform="translate(-1.125 1.125)" />
		</clipPath>
	</defs>
</svg>

```

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:1]: Nitpick: I think there should be some margin between the radio buttons. Currently, they appear very close together. What do you think?

image

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx]: I think these messages are too specific to the "Related Products" collection. For other collections like "Upsells", the wording wouldn't be accurate. For example, with Upsells, it would be "Upsells will be pulled from the product a shopper is currently viewing."

To make this more versatile, I suggest we use a more generic copy that works for all collection types. What do you think?

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx]:
```suggestion
const enum PRODUCT_REFERENCE_TYPE {
```

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx]:
```suggestion
	const REFERENCE_TYPE_PRODUCT = 'product';
```

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx]:
```suggestion
	const showSpecificProductSelector = showRadioControl
```

**Review by imanish003 — CHANGES_REQUESTED** [review]
Hey Karol! Great job. The implementation looks solid and everything seems to be working as expected 🙌🏻
- ✅ "From the current product" is selected by default
- ✅ Switching between "From the current product" and "From a specific product" keeps the correct product selected
- ✅ In new post, "From the current product" isn't visible, as expected

I left some nitpicks. There's one concern regarding the copy text, which I believe needs to be addressed before merging. Let me know your thoughts on that. I am happy to approve the PR after that 🚀

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (octaedro) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:45]: Can we also check whether the image source is `undefined` to show the placeholder? Otherwise,  when the product doesn't have an image will look like this

Screenshot 2024-09-17 at 10 39 03 AM

Comment (octaedro) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:45]: If we add a product here that we send later to the trash can, should we continue showing it?

**Review by octaedro — COMMENTED** [review]
Good job @kmanijak! This is testing well on my end and the code LGTM. I just left a couple of small questions.

File: plugins/woocommerce/assets/images/blocks/product-collection/placeholder.svg
Comment (kmanijak) [reply plugins/woocommerce/assets/images/blocks/product-collection/placeholder.svg]: > I really like how this type of icons are handled in [new Product Editor](https://github.com/woocommerce/woocommerce/blob/add/50916-product-reference-updated/packages/js/product-editor/src/images) (See [storybook](https://woocommerce.github.io/woocommerce/?path=/story/product-editor-images-cashregister--default)).

I see. I like the approach but I think we should follow what the rest of Core and Blocks is doing which is keeping the images in assets.

> Although, we should consider formatting this SVG icon for improved readability 🙂

Sure thing! In addition, I ran the SVG through https://jakearchibald.github.io/svgomg/ and optimized it a bit.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:1]: Good catch, I agree. I increased the gap from default 4px to 8px. Looks much cleaner!
Without focus | With focus
-- | --
image | image

(Unless you think we should increase more but it's starting to look off then).

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx]: Agree! I changed to:

- 'Linked products will be pulled from the product a shopper is currently viewing'
- 'Select a product to pull the linked products from'

In future iterations, we can make it more contextual. Or actually, it could part of collections PR (Upsells, Related Products) to modify these. Since the collections aren't yet merged I don't want to include that in this PR 🙌

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:45]: > Can we also check whether the image source is undefined to show the placeholder? Otherwise, when the product doesn't have an image will look like this

Oh, nice catch! ❤️

Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:45]: > If we add a product here that we send later to the trash can, should we continue showing it?

Good thinking! We have it covered in the upper scope. So basically if the product is not available, merchant won't even get to this state but will be prompted to choose another product earlier.

It was Implemented in this PR https://github.com/woocommerce/woocommerce/pull/51114 and here's how it looks like, the promp shows up instead of the Product Collection block:

image

**Comment by kmanijak** [comment]
@imanish003 , @octaedro thanks for thorough reviews! I addressed all of your comments so I'd appreciate another round 🙌

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:1]: Looks good to me. Thanks for changing it 🚀

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx]: Sounds good to me. Thanks for making changes 🚀

**Review by imanish003 — COMMENTED** [review]
Hey @kmanijak!

Thanks for addressing my comments. One thing I noticed is that even when I choose "From a specific product", preview label is still showing. IMO preview label should be hidden when we know the exact product reference. What do you think?

Apart from this, everything else look good to me 🚀

**Review by imanish003 — APPROVED** [review]

**Comment by imanish003** [comment]
I've created a new issue to track hiding the preview label when "From a specific product" is selected: #51601.

I am going to merge this PR now as we can address this in a follow-up PR. Great work on this @kmanijak  🚀

**Pull request merged**
